functions/access.js

```diff
@@ -52,65 +60,95 @@ const getAccessDoc = async (venueId, method) => {
+
+  if (!access || !access.exists || !access.data().emails) {
     return false;
   }
+
   return access.data().emails.includes(email.trim().toLowerCase());
 };
 
 const isValidCode = async (venueId, code) => {
+  if (!venueId || !code) return false;
+
   const access = getAccessDoc(venueId, "code");
-  if (!access.exists || !access.data().codes) {
+
+  if (!access || !access.exists || !access.data().codes) {
     return false;
   }
+
   return access.data().codes.includes(code.trim());
 };
 
 const createToken = async (venueId, uid, password, email, code) => {
-  if (!venueId || !uid || !password || !email || !code) return {};
+  if (!venueId || !uid || !password || !email || !code) return undefined;
 
-  const venue = await admin.firestore().collection("venues").doc(venueId).get();
-  if (!venue.exists) {
+  const venueRef = admin.firestore().collection("venues").doc(venueId);
+  const accessRef = admin.firestore().collection("accessgranted").doc(uid);
+
+  if (!venueRef.exists) {
     throw new HttpsError("not-found", `venue ${venueId} does not exist`);
   }
 
-  const token = uuid();
-  const grantedDoc = await venue.ref.collection("accessgranted").doc(uid).get();
-
-  const tokenData = {
-    usedAt: [Date.now()],
-    password: password || null,
-    email: email || null,
-    code: code || null,
-  };
-
-  if (grantedDoc.exists) {
-    await grantedDoc.update({ [`tokens.${token}`]: tokenData });
-  } else {
-    const granted = { tokens: { [token]: tokenData } };
-    await grantedDoc.set(granted);
-  }
-  return { token };
+  return await admin
+    .firestore()
+    .runTransaction(async (transaction) => {
+      const [venue, granted] = await Promise.all([
+        transaction.get(venueRef),
+        transaction.get(accessRef),
+      ]);
+
+      if (!venue.exists) {
+        return undefined;
+      }
+
+      const token = uuid();
+
+      const tokenData = {
+        usedAt: [Date.now()],
+        password: password,
+        email: email,
+        code: code,
+      };
+
+      const newToken = { [token]: tokenData };
+      if (granted.exists) {
+        transaction.update(accessRef, newToken);
+      } else {
+        transaction.set(accessRef, newToken);
+      }
+
+      return token;
+    })
+    .catch((e) => {
+      console.log("Transaction failure:", e);
+      return undefined;
+    });
 };
 
 exports.checkAccess = functions.https.onCall(async (data, context) => {
-  console.log("checkAccess", data, context.auth);
-  const isValidToken =
-    context &&
-    context.auth &&
-    context.auth.uid &&
-    (await checkIsValidToken(data.venueId, context.auth.uid, data.token));
-  if (isValidToken) {
-    console.log(`valid token, returning ${data.token}`);
+  if (!data || !context) return { token: undefined };
+  const isValidToken = await checkIsValidToken(
+    data.venueId,
+    context.auth.uid,
+    data.token
+  );
+  if (context && context.auth && context.auth.uid && isValidToken) {
```

Comment:
The `context.auth.uid` could crash if any of the 'path' aren't defined, since the `if` check has been moved below it now.

Reply:
I believe my previous suggestion was just collapsing the entire `const` into the `if` statement so that there is no longer a `isValidToken` `const` at all.

Reply:
Applied, your comment is valid.

Reply:
Awesome :)